tests/system_tests_three_routers.py

```diff
@@ -624,5 +636,151 @@ def run(self):
+        self.far_messages  = 0
+
+
+    def timeout ( self ):
+        self.check_results ( )
+        self.bail ( "Timeout Expired " )
+
+
+    def bail ( self, text ):
+        self.timer.cancel()
+        self.error = text
+        self.send_cnx.close()
+        self.near_cnx.close()
+        self.mid_cnx.close()
+        self.far_cnx.close()
+
+
+    def on_start ( self, event ):
+        self.timer       = event.reactor.schedule  ( TIMEOUT, Timeout(self) )
+        self.send_cnx    = event.container.connect ( self.near_router )
+        self.near_cnx    = event.container.connect ( self.near_router )
+        self.mid_cnx     = event.container.connect ( self.mid_router )
+        self.far_cnx     = event.container.connect ( self.far_router )
+
+        self.sender      = event.container.create_sender   ( self.send_cnx, self.dest)
+
+        self.near_recv_1 = event.container.create_receiver  ( self.near_cnx, self.dest)
+        self.near_recv_2 = event.container.create_receiver  ( self.near_cnx, self.dest)
+
+        self.mid_recv_1  = event.container.create_receiver  ( self.mid_cnx,  self.dest)
+        self.mid_recv_2  = event.container.create_receiver  ( self.mid_cnx,  self.dest)
+
+        self.far_recv_1  = event.container.create_receiver  ( self.far_cnx,  self.dest)
+        self.far_recv_2  = event.container.create_receiver  ( self.far_cnx,  self.dest)
+
+        self.near_recv_1.flow ( self.n_expected )
+        self.mid_recv_1.flow  ( self.n_expected )
+        self.far_recv_1.flow  ( self.n_expected )
+
+
+    def on_sendable ( self, event ):
+        msg = Message ( body     = "Hello, closest.",
+                        address  = self.dest
+                      )
+        event.sender.send ( msg )
+
+
+    def on_message ( self, event ):
+
+        self.n_received += 1
+
+        if self.n_received == 10:
+            # The first ten messages should have gone exclusively
+            # to the near receivers.  At this point we should have
+            # no messages in the mid or far receivers.
+            self.near_recv_1.close()
+            self.near_recv_2.close()
+            if self.debug :
+                self.mid_messages = 1
+            if self.mid_messages > 0 or self.far_messages > 0 :
+                self.bail ( "error: mid or far receivers got messages before near were closed." )
+        elif self.n_received == 20:
+            # The next 10 messages should have gone exclusively
+            # to the mid receivers.  At this point we should have
+            # no messages in the far receivers.
+            self.mid_recv_1.close()
+            self.mid_recv_2.close()
+            if self.far_messages > 0 :
+                self.bail ( "error: far receivers got messages before mid were closed." )
+
+        # Increment the near, mid, or far counts, depending on
+        # which receiver the message came in on.
+        if event.receiver   == self.near_recv_1 or event.receiver == self.near_recv_2:
```

Comment:
No messages are ever received by self.near_recv_2 because of lack for credit. Same with self.mid_recv_2 and self.far_recv_2. These 3 receivers should also send some flow so that they also receive message